File: Silver/Training/money.py
"""
ID: chenant1
LANG: PYTHON3
TASK: money
"""

def find_solutions(coins, value): #DYNAMIC PROGRAMMING
  solutions = [0 for i in range(value + 1)]
  solutions[0] = 1
  if coins[0] == 1:
    solutions[1] = 1
  else:
    solutions[1] = 0
  for j in range(len(coins)):
    for i in range(2, value + 1):
      if coins[j] <= i:
        solutions[i] += solutions[i - coins[j]]
  print(solutions)
  return solutions[-1]

with open("money.in", "r") as fin, open("money.out", "w") as fout:
  N = int(fin.readline().split()[1])
  coins = []
  for i in fin.readlines():
    coins += [int(j) for j in i.split()]
  coins = sorted(coins)
  print(coins)
  fout.write(str(find_solutions(coins, N)) + "\n")

File: Silver/Training/money.cc
/*
ID: chenant1
LANG: C
TASK: money
*/

#include <stdio.h>
#define getint(i) scanf("%d", &i)

long long f[10001];

int main(void)
{
	int n, m;
	int i, j, k, t;
	freopen("money.in", "r", stdin);
	freopen("money.out", "w", stdout);
	getint(n);
	getint(m);
	f[0] = 1;
	for(i = 1; i <= n; i++){
		getint(t);
		for(j = t; j <= m; j++){
			f[j] += f[j - t];
		}
	}
	printf("%lld\n", f[m]);
	return 0;
}